hugr-core/src/ops/constant/custom.rs

```diff
@@ -90,6 +88,33 @@ pub trait CustomConst:
     fn get_type(&self) -> Type;
 }
 
+/// Prerequisite for `CustomConst`. Allows to declare a custom hash function, but the easiest
+/// options are either to `impl MaybeHash for ... {}` to declare "not hashable", or else
+/// to implement (or derive) [Hash].
+pub trait MaybeHash {
+    /// Hashes the value, if possible; else return `false` without mutating the `Hasher`.
+    /// This relates with [CustomConst::equal_consts] just like [Hash] with [Eq]:
+    /// * if `x.equal_consts(y)` ==> `x.maybe_hash(s)` behaves equivalently to `y.maybe_hash(s)`
+    /// * if `x.hash(s)` behaves differently from `y.hash(s)` ==> `x.equal_consts(y) == false`
+    ///
+    /// As with [Hash], these requirements can trivially be satisfied by either
+    /// * `equal_consts` always returning `false`, or
+    /// * `maybe_hash` always behaving the same (e.g. returning `false`, as it does by default)
+    ///
+    /// Note: this uses `dyn` rather than being parametrized by `<H: Hasher>` so that we can
+    /// still use `dyn CustomConst`.
+    fn maybe_hash(&self, _state: &mut dyn Hasher) -> bool {
+        false
+    }
+}
+
+impl<T: Hash> MaybeHash for T {
+    fn maybe_hash(&self, st: &mut dyn Hasher) -> bool {
+        Hash::hash(self, &mut Box::new(st));
```

Comment:
```suggestion
    fn maybe_hash(&self, mut st: &mut dyn Hasher) -> bool {
        Hash::hash(self, &mut st);
```

Reply:
That gives me:
```
error[E0277]: the size for values of type `dyn Hasher` cannot be known at compilation time
   --> hugr-core/src/ops/constant/custom.rs:113:26
    |
113 |         Hash::hash(self, &mut *st);
    |         ----------       ^^^^^^^^ doesn't have a size known at compile-time
    |         |
    |         required by a bound introduced by this call
    |
    = help: the trait `Sized` is not implemented for `dyn Hasher`
```
which I admit is odd - I'm not trying to pass a Hasher, only a `&mut` to it. Am I missing something?

Reply:
Ah sorry I'd put in an extra `*`

Reply:
`*st` is trying to dereference an unsized value into the stack.
The suggestion avoids that by using `T : Hash` $\implies$ &mut T : Hash`

Reply:
This does look to me as if I'm passing a mutable reference to a mutable reference!